### PR TITLE
LwM2M Bootstrap support fixes

### LINKED ISSUE #18080: LWM2M bootstrap issue
I'm not sure if the issue I describe here is a bug report or my misunderstanding, hence not marking it as a bug yet. The documentation is quite scarce on how to initiate the bootstrap with the LWM2M library so I might miss some information.

I've tried to run the LWM2M bootstrap in the `lwm2m_client` sample, using local LWM2M bootstrap server (`leshan-bsserver-demo.jar` from the same source as the `leshan-server-demo.jar` from docs). What I did (all w/o security enabled):
1.  Started the local BS server instance at port 5683, along with local LWM2M server instance at port 5783.
2. Configured the BS server with client endpoint name (`qemu_x86`) and local LWM2M server address (`coap://localhost:5783`)
3. Enabled bootstrap support in the sample (`CONFIG_LWM2M_RD_CLIENT_SUPPORT_BOOTSTRAP`) and increased the maximum number of security and server object instances.
4. Configured the security object, to indicate it points to a bootstrap server (`lwm2m_engine_set_u8("0/0/1", 1);`)

After this, I was able to initiate the LWM2M bootstrap procedure, yet the client was not able to finalize it.  It seems, that the procedure that should set the appropriate state in the state machine (`engine_bootstrap_finish()`) is not called anywhere.  When the server sent the Bootstrap-finish message, the client replied with 4.04 Not Found. This clearly indicates that I miss a handler for the `Bootstrap-finish` message. 

image

So my question is, is the `Bootstrap-finish` handler supposed to be present in LWM2M engine and we miss it or is the application supposed to provide it and register it somehow with LWM2M on its own?

Also, I have a side question regarding client deregistration - is it possible to deregister a LWM2M server with current implementation? I saw there is an `ENGINE_DEREGISTER` state which seems to handle the deregistration procedure, yet I've seen no way to set this state. Is it possible to trigger deregistration from the application somehow?

### PR DESCRIPTION
**Pull request by mike-scott**
When bootstrap support is enabled:
- add second instances for server and security objects
- support bootstrap-finish notification from server (POST to /bs)
- mark first server as a bootstrap server in the sample

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/18080

### CODE REVIEW

**Comment by ioannisg** [comment]
@tbursztyka could you provide some feedback here? Ideally, this should go into 2.0

**Review by tbursztyka — APPROVED** [review]

File: samples/net/lwm2m_client/src/lwm2m-client.c [hunk 1]
```diff
@@ -282,6 +282,16 @@ static int lwm2m_setup(void)
 				(void *)client_psk, sizeof(client_psk));
 #endif /* CONFIG_LWM2M_DTLS_SUPPORT */
 
+#if defined(CONFIG_LWM2M_RD_CLIENT_SUPPORT_BOOTSTRAP)
+	lwm2m_engine_set_u8("0/0/1", 1);
+
+	/* Create 2nd instances of server and security objects needed for
```
Comment (ioannisg) [inline samples/net/lwm2m_client/src/lwm2m-client.c]: 2nd instance?

Comment (ioannisg) [reply samples/net/lwm2m_client/src/lwm2m-client.c]: or two instances?

**Review by ioannisg — APPROVED** [review]
Minor typo, otherwise looks ok to me

**Comment by mike-scott** [comment]
Cleaned up the comments.  Here's a diff:
```
diff --git a/samples/net/lwm2m_client/src/lwm2m-client.c b/samples/net/lwm2m_client/src/lwm2m-client.c
index a540c00a97..84052c8f5b 100644
--- a/samples/net/lwm2m_client/src/lwm2m-client.c
+++ b/samples/net/lwm2m_client/src/lwm2m-client.c
@@ -283,10 +283,11 @@ static int lwm2m_setup(void)
 #endif /* CONFIG_LWM2M_DTLS_SUPPORT */
 
 #if defined(CONFIG_LWM2M_RD_CLIENT_SUPPORT_BOOTSTRAP)
+	/* Mark 1st instance of security object as a bootstrap server */
 	lwm2m_engine_set_u8("0/0/1", 1);
 
-	/* Create 2nd instances of server and security objects needed for
-	 * bootstrap
+	/* Create 2nd instance of server and security objects needed for
+	 * bootstrap process
 	 */
 	lwm2m_engine_create_obj_inst("0/1");
 	lwm2m_engine_create_obj_inst("1/1");
```

**Pull request merged**